framework/src/modules/fee/method.ts

```diff
@@ -12,6 +12,51 @@
  * Removal or modification of this copyright notice is prohibited.
  */
 
+import { getContextStoreBigInt, getContextStoreBool, MethodContext } from '../../state_machine';
 import { BaseMethod } from '../base_method';
+import { CONTEXT_STORE_KEY_CCM_PROCESSING } from '../interoperability/constants';
+import { CONTEXT_STORE_KEY_AVAILABLE_CCM_FEE, CONTEXT_STORE_KEY_AVAILABLE_FEE } from './constants';
+import { InsufficientFeeEvent } from './events/insufficient_fee';
+import { ModuleConfig } from './types';
 
-export class FeeMethod extends BaseMethod {}
+export class FeeMethod extends BaseMethod {
+	private _config!: ModuleConfig;
+
+	public init(config: ModuleConfig) {
+		this._config = config;
+	}
+
+	public getFeeTokenID(): Buffer {
+		return this._config.feeTokenID;
+	}
+
+	public payFee(methodContext: MethodContext, amount: bigint): void {
+		const isCCMProcessing = getContextStoreBool(
+			methodContext.contextStore,
+			CONTEXT_STORE_KEY_CCM_PROCESSING,
+		);
+		if (isCCMProcessing) {
+			const availableFee = getContextStoreBigInt(
+				methodContext.contextStore,
+				CONTEXT_STORE_KEY_AVAILABLE_CCM_FEE,
+			);
+			if (availableFee < amount) {
+				this.events.get(InsufficientFeeEvent).error(methodContext);
+				methodContext.contextStore.delete(CONTEXT_STORE_KEY_AVAILABLE_CCM_FEE);
+				throw new Error('Cross-chain message ran out of fee.');
+			}
+			methodContext.contextStore.set(CONTEXT_STORE_KEY_AVAILABLE_CCM_FEE, availableFee - amount);
+			return;
+		}
+		const availableFee = getContextStoreBigInt(
+			methodContext.contextStore,
+			CONTEXT_STORE_KEY_AVAILABLE_FEE,
+		);
+		if (availableFee < amount) {
+			this.events.get(InsufficientFeeEvent).error(methodContext);
+			methodContext.contextStore.delete(CONTEXT_STORE_KEY_AVAILABLE_FEE);
+			throw new Error('Transaction ran out of fee.');
+		}
+		methodContext.contextStore.set(CONTEXT_STORE_KEY_AVAILABLE_FEE, availableFee - amount);
```

Comment:
I don't see a good reason to duplicate this code. Such things tend to increase maintenance complexity and lead to inconsistency.

This functionality could be generalized to a private method e.g:

```typescript
private updateContextStore(isCCMProcessing: boolean): void
```

[not sure that it is the most appropriate name, but it's just to illustrate the idea]

Based on the input argument, it would set appropriate variables for:

1. the error message to be thrown `outOfFeeError`
1. the key for the context store: `contextStoreKey`

Those seem to be the only 2 differences between the two cases.

Then move the whole generalized functionality to the private method:

```typescript
methodContext.contextStore.delete(contextStoreKey);
throw new Error(outOfFeeError);

...

methodContext.contextStore.set(contextStoreKey, availableFee - amount);
```

Reply:
What I had in mind was setting the values for `key` and `outOfFeeMsg` variables inside `if` section, and then calling `_deductFeeFromKey()` with those variables as arguments after `if`, but this is OK, too 🙂